File: src/packages/control-ui/src/api/quic_evasion_tunnel_coordinator.ts
import {
  SegmentationStrategy,
  SniSegmentationMasquerader,
} from './sni_segmentation_masquerader.js';
import { QuicConnectionController, QuicConnectionState } from './quic_connection_controller.js';
import { QuicHeaderType, QuicPacketCodec } from './quic_packet_codec.js';
import type { QuicPacketHeader } from './quic_packet_codec.js';
import { QuicStreamMultiplexer, QuicStreamType } from './quic_stream_multiplexer.js';
import type { QuicStreamFrame } from './quic_stream_multiplexer.js';
import { ReplayResistantTunnelSession } from './replay_resistant_tunnel.js';
import type { ReplayResistantConfig } from './replay_resistant_tunnel.js';

export interface QuicTunnelMetrics {
  connectionState: QuicConnectionState;
  totalDatagramsSent: number;
  totalDatagramsReceived: number;
  activeStreams: number;
  evasionActive: boolean;
}

export class QuicEvasionTunnelCoordinator {
  private multiplexer: QuicStreamMultiplexer;
  private connection: QuicConnectionController;
  private masquerader: SniSegmentationMasquerader;
  private replaySession: ReplayResistantTunnelSession;
  public destCid: Uint8Array;
  public srcCid: Uint8Array;
  private nextPacketNum: number = 1;
  private totalSent: number = 0;
  private totalRecv: number = 0;
  public evasionMode: boolean;

  constructor(
    destCid: Uint8Array,
    srcCid: Uint8Array,
    clientRandom: Uint8Array,
    serverRandom: Uint8Array,
    evasionMode: boolean = false,
  ) {
    const replayConfig: ReplayResistantConfig = {
      psk: new Uint8Array(32).fill(0x3a),
      maxSkewSecs: 30,
      maxTrackedNonces: 10000,
    };
    this.replaySession = new ReplayResistantTunnelSession(replayConfig, clientRandom, serverRandom);

    this.connection = new QuicConnectionController(65536);
    this.connection.setState(QuicConnectionState.Established);

    this.multiplexer = new QuicStreamMultiplexer(65536);
    this.masquerader = new SniSegmentationMasquerader(SegmentationStrategy.MidSniSplit, 8, 32);

    this.destCid = destCid;
    this.srcCid = srcCid;
    this.evasionMode = evasionMode;
  }

  openTunnelStream(): number {
    return this.multiplexer.openStream(QuicStreamType.ClientBidirectional);
  }

  prepareOutboundDatagram(
    streamId: number,
    appData: Uint8Array,
    timestampSecs: number,
  ): Uint8Array[] {
    // 1. Multiplex stream frame
    const streamFrame = this.multiplexer.writeStreamData(streamId, appData, false);
    // Serialize frame as JSON string in Uint8Array
    const frameObj = {
      streamId: streamFrame.streamId,
      offset: streamFrame.offset,
      fin: streamFrame.fin,
      data: Array.from(streamFrame.data),
    };
    const frameBytes = new TextEncoder().encode(JSON.stringify(frameObj));

    // 2. Wrap in QUIC 1-RTT Short packet
    const header: QuicPacketHeader = {
      headerType: QuicHeaderType.OneRttShort,
      version: 0,
      destCid: this.destCid,
      srcCid: this.srcCid,
      packetNumber: this.nextPacketNum,
    };
    const quicPacket = QuicPacketCodec.encodePacket(header, frameBytes);

    // 3. Seal with replay resistance
    const sealed = this.replaySession.sealPacket(this.nextPacketNum, timestampSecs, quicPacket);

    this.connection.onPacketSent(this.nextPacketNum, sealed.length, timestampSecs * 1000);
    this.nextPacketNum += 1;
    this.totalSent += 1;

    // 4. If evasion mode is enabled, segment datagram to evade DPI
    if (this.evasionMode) {
      return this.masquerader.segmentStream(sealed, this.nextPacketNum);
    } else {
      return [sealed];
    }
  }

  processInboundDatagram(
    sealedDatagram: Uint8Array,
    currentTimeSecs: number,
  ): { streamId: number; payload: Uint8Array } {
    // 1. Verify replay resistance and open envelope
    const { sequence, payload: quicPacket } = this.replaySession.openPacket(
      sealedDatagram,
      currentTimeSecs,
    );

    this.connection.onAckReceived(sequence, currentTimeSecs * 1000);
    this.totalRecv += 1;

    // 2. Decode QUIC packet
    const { payload } = QuicPacketCodec.decodePacket(quicPacket, this.destCid.length);

    // 3. Deserialize stream frame
    const jsonStr = new TextDecoder().decode(payload);
    const parsed = JSON.parse(jsonStr);
    const frame: QuicStreamFrame = {
      streamId: parsed.streamId,
      offset: parsed.offset,
      fin: parsed.fin,
      data: new Uint8Array(parsed.data),
    };

    // 4. Reassemble stream data
    const assembled = this.multiplexer.receiveStreamFrame(frame);
    return { streamId: frame.streamId, payload: assembled };
  }

  getTunnelMetrics(): QuicTunnelMetrics {
    return {
      connectionState: this.connection.getMetrics().state,
      totalDatagramsSent: this.totalSent,
      totalDatagramsReceived: this.totalRecv,
      activeStreams: 1,
      evasionActive: this.evasionMode,
    };
  }
}

File: labs/mobile/donor-archive/tunnel/QuicEvasionTunnelCoordinator.kt
package com.luminet.android.tunnel

data class QuicTunnelMetrics(
    val connectionState: QuicConnectionState,
    val totalDatagramsSent: Long,
    val totalDatagramsReceived: Long,
    val activeStreams: Int,
    val evasionActive: Boolean
)

class QuicEvasionTunnelCoordinator(
    val destCid: ByteArray,
    val srcCid: ByteArray,
    clientRandom: ByteArray,
    serverRandom: ByteArray,
    val evasionActive: Boolean
) {
    private val multiplexer = QuicStreamMultiplexer(65536)
    private val connection = QuicConnectionController(65536).apply {
        state = QuicConnectionState.ESTABLISHED
    }
    private val masquerader = SniSegmentationMasquerader(SniSegmentationStrategy.MID_SNI_SPLIT, 8, 32)
    private val replaySession = ReplayResistantTunnelSession(
        ReplayResistantConfig(),
        clientRandom,
        serverRandom
    )
    private val codec = QuicPacketCodec()
    private var nextPacketNum: Long = 1
    private var totalSent: Long = 0
    private var totalRecv: Long = 0

    fun openTunnelStream(): Long {
        return multiplexer.openStream(QuicStreamType.CLIENT_BIDIRECTIONAL)
    }

    fun prepareOutboundDatagram(streamId: Long, appData: ByteArray, timestampSecs: Long): List<ByteArray> {
        val streamFrame = multiplexer.writeStreamData(streamId, appData, false)
        // Serialize frame payload simply: [streamId: 8][offset: 8][fin: 1][data]
        val frameBytes = java.nio.ByteBuffer.allocate(17 + streamFrame.data.size).apply {
            putLong(streamFrame.streamId)
            putLong(streamFrame.offset)
            put(if (streamFrame.fin) 1.toByte() else 0.toByte())
            put(streamFrame.data)
        }.array()

        val hdr = QuicPacketHeader(
            headerType = QuicHeaderType.ONE_RTT_SHORT,
            version = 0,
            destCid = destCid,
            srcCid = srcCid,
            packetNumber = nextPacketNum
        )
        val quicPkt = codec.encodePacket(hdr, frameBytes)

        val sealed = replaySession.sealPacket(nextPacketNum, timestampSecs, quicPkt)
        connection.onPacketSent(nextPacketNum, sealed.size, timestampSecs * 1000)

        nextPacketNum++
        totalSent++

        return if (evasionActive) {
            masquerader.segmentStream(sealed, nextPacketNum)
        } else {
            listOf(sealed)
        }
    }

    fun processInboundDatagram(sealedDatagram: ByteArray, currentTimeSecs: Long): Pair<Long, ByteArray> {
        val (seq, _, quicPkt) = replaySession.openPacket(sealedDatagram, currentTimeSecs)
        connection.onAckReceived(seq, currentTimeSecs * 1000)
        totalRecv++

        val (_, payload) = codec.decodePacket(quicPkt, destCid.size)
        val buf = java.nio.ByteBuffer.wrap(payload)
        val streamId = buf.long
        val offset = buf.long
        val fin = buf.get() == 1.toByte()
        val data = ByteArray(payload.size - 17)
        buf.get(data)

        val frame = QuicStreamFrame(streamId, offset, fin, data)
        val assembled = multiplexer.receiveStreamFrame(frame)

        return Pair(streamId, assembled)
    }

    fun getTunnelMetrics(): QuicTunnelMetrics {
        return QuicTunnelMetrics(
            connectionState = connection.state,
            totalDatagramsSent = totalSent,
            totalDatagramsReceived = totalRecv,
            activeStreams = 1,
            evasionActive = evasionActive
        )
    }
}
